CNV-54462 -- vTPM documentation (#96528)
openshift/openshift-docs: modules/virt-about-vtpm-devices.adoc

If you do not enable vTPM, then the VM does not recognize a TPM device, even if
the node has one.

A vTPM device also protects virtual machines by storing secrets without physical hardware. {VirtProductName} supports persisting vTPM device state by using Persistent Volume Claims (PVCs) for VMs. You must specify the storage class to be used by the PVC by setting the `vmStateStorageClass` attribute in the `HyperConverged` custom resource (CR).

In {VirtProductName}, if the `storageClass` for `vmStateStorageClass` is not explicitly defined, the system uses the cluster’s default storage class to create the vTPM PVC. The access mode for the PVC is determined by the storage profile of the storage class.

[IMPORTANT]
====
**vTPM storage class and live migration considerations**

> VMs with vTPM devices that use a ReadWriteOnce (RWO) volume are non-migratable. To enable live migration, the underlying storage class for the vTPM PVC must support the ReadWriteMany (RWX) access mode.

Reviewer: Looks much better! But currently, this bolded text title on line 25 is rendering on the same line as the rest of the admonition contents. To fix that, you'd add a blank line between them, but I searched the repo and don't see any other instances of adding a title to an admonition. To be on the safe side, please remove `**vTPM storage class and live migration considerations**`. Then it should be good to merge.

Author: Thanks for all these suggestions, Pan! I really appreciate your help :)

A vTPM device also protects virtual machines by storing secrets without physical hardware. {VirtProductName} supports persisting vTPM device state by using Persistent Volume Claims (PVCs) for VMs. You must specify the storage class to be used by the PVC by setting the `vmStateStorageClass` attribute in the `HyperConverged` custom resource (CR):

> ====

Author: Thanks for pointing this out. I've added it under IMPORTANT.

**vTPM Storage Class and Live Migration Considerations**
In OpenShift Virtualization 4.17, if the `storageClass` for `vmStateStorageClass` is not explicitly defined, the system uses the cluster’s default storage class to create the vTPM persistent volume claim (PVC). The access mode for the PVC is determined by the storage profile of the storage class and is not always ReadWriteOnce (RWO).

In 4.17, VMs with vTPM devices that use an RWO volume are non-migratable. To enable live migration, the underlying storage class for the vTPM PVC must support the ReadWriteMany (RWX) access mode.

> From OpenShift Virtualization 4.18, live migration is supported for vTPM-enabled VMs even when the vTPM PVC uses the RWO access mode.

Reviewer 2: i'm not sure if the 4.18 information is needed here. since this is 4.17 documentation i'm unclear if it's normal for us to mention what would be in a future release. can you let me know if it was specifically requested to be put in? then i'll ask a fellow reviewer their opinion

Author: After confirming with the QA @nijinashok, I've removed the line.

> A vTPM device also protects virtual machines by storing secrets without physical hardware. {VirtProductName} supports persisting vTPM device state by using Persistent Volume Claims (PVCs) for VMs. You must specify the storage class to be used by the PVC by setting the `vmStateStorageClass` attribute in the `HyperConverged` custom resource (CR):

Reviewer: I would change the colon at the end of this paragraph to a period, since you're adding more info in between this para and the code block.

[IMPORTANT]
====

> **vTPM Storage class and live migration considerations**

Reviewer: ```suggestion
**vTPM storage class and live migration considerations**
```

> In {VirtProductName}, if the `storageClass` for `vmStateStorageClass` is not explicitly defined, the system uses the cluster’s default storage class to create the vTPM persistent volume claim (PVC). The access mode for the PVC is determined by the storage profile of the storage class.

Reviewer: PVC was already defined earlier in the module:

```suggestion
In {VirtProductName}, if the `storageClass` for `vmStateStorageClass` is not explicitly defined, the system uses the cluster’s default storage class to create the vTPM PVC. The access mode for the PVC is determined by the storage profile of the storage class.
```

Reviewer: It occurs to me that the second paragraph of this admonition is the really important part and we may be distracting from it by including the other information upfront. I would suggest taking the first paragraph out and putting it in the normal body text.

> VMs with vTPM devices that use an RWO volume are non-migratable. To enable live migration, the underlying storage class for the vTPM PVC must support the ReadWriteMany (RWX) access mode.

Reviewer: ```suggestion
VMs with vTPM devices that use a ReadWriteOnce (RWO) volume are non-migratable. To enable live migration, the underlying storage class for the vTPM PVC must support the ReadWriteMany (RWX) access mode.
```

[NOTE]
====
The storage class must be of type `Filesystem` and support the `ReadWriteMany` (RWX) access mode.

> ====

Reviewer: Maybe we should remove this note in favor of the IMPORTANT block. (You can add the `Filesystem` detail to that block. I'd also double check with an SME to make sure it all still makes sense.)

Author: Thanks for the suggestion, Pan. For now, I’d prefer to keep the NOTE since I don’t see a requirement to remove it and it still provides relevant context. Could you please clarify what part you think is redundant compared to the IMPORTANT block? Happy to update if we get an SME confirmation.

Reviewer: No worries. The note restates the RWX requirement (without including the reason). But the simpler note might be easier for someone to see. I think it's fine to leave it for now.
